frontend/dockerfile/docs/syntax.md

```diff
@@ -237,3 +237,91 @@ RUN --network=none pip install --find-links wheels mypackage
 
 `pip` will only be able to install the packages provided in the tarfile, which
 can be controlled by an earlier build stage.
+
+
+## Here-Documents
+
+To use this flag, set Dockerfile version to `labs` channel. Currently this feature is only available
+in `docker/dockerfile-upstream:master-labs` image.
+
+```
+# syntax=docker/dockerfile-upstream:master-labs
+```
+
+Here-documents allow redirection of subsequent Dockerfile lines to the input of `RUN` or `COPY` commands.
+If such command contains a [here-document](https://pubs.opengroup.org/onlinepubs/9699919799/utilities/V3_chap02.html#tag_18_07_04)
+Dockerfile will consider the next lines until the line only containing a here-doc delimiter as part of the same command.
+
+#### Example: running a multi-line script
+
+```dockerfile
+# syntax = docker/dockerfile-upstream:master-labs
+FROM debian
+RUN <<eot bash
+  apt-get update
+  apt-get install -y vim
+eot
+```
+
+If the command only contains a here-document, its contents is evaluated with default shell.
+
+```dockerfile
+# syntax = docker/dockerfile:master-labs
+FROM debian
+RUN <<eot
+  mkdir -p foo/bar
+eot
+```
+
+Alternatively, shebang header can be used to define an interpreter.
+
+```dockerfile
+# syntax = docker/dockerfile:master-labs
+FROM python:3.6
+RUN <<eot
+#!/usr/bin/env python
+print("hello world")
+eot
+```
+
+More complex examples may use multiple here-documents.
+
+```dockerfile
+# syntax = docker/dockerfile:master-labs
+FROM alpine
+RUN <<eot cat > file1 && <<eot2 cat > file2
+I am
+first
+eot
+I am
+second
+eot2
+```
+
+#### Example: creating inline files
+
+In `COPY` commands source parameters can be replaced with here-doc indicators.
+Regular here-doc variable expansion and tab stripping rules apply.
+
+```dockerfile
+# syntax = docker/dockerfile:master-labs
+FROM alpine
+ARG FOO=bar
+RUN <<-eot /app/foo
+	hello ${FOO}
+eot
+```
+
+```dockerfile
```

Comment:
Added another link to here-docs docs. I don't want to copy everything from there, otherwise, we should copy the whole shell documentation for the `RUN` command.